arangod/Cluster/RebootTracker.h

```diff
@@ -23,119 +23,86 @@
-
-  class PeerState {
-   public:
-    PeerState(ServerID serverId, RebootId rebootId)
-        : _serverId{std::move(serverId)}, _rebootId{rebootId} {}
-
-    ServerID const& serverId() const noexcept { return _serverId; }
-    RebootId rebootId() const noexcept { return _rebootId; }
-
-    void toVelocyPack(velocypack::Builder& builder) const;
-    static PeerState fromVelocyPack(velocypack::Slice);
-
-   private:
-    ServerID _serverId;
-    RebootId _rebootId;
-  };
-
-  explicit RebootTracker(SchedulerPointer scheduler);
-
-  CallbackGuard callMeOnChange(PeerState const& peerState, Callback callback,
-                               std::string callbackDescription);
-
-  void updateServerState(
-      containers::FlatHashMap<ServerID, RebootId> const& state);
-
- private:
-  using CallbackId = uint64_t;
-
+  using Callback = fu2::unique_function<void()>;
   struct DescriptedCallback {
     Callback callback;
     std::string description;
   };
+  using State = containers::FlatHashMap<ServerID, RebootId>;
+  using CallbackId = uint64_t;
+  using RebootIds =
+      std::map<RebootId,
+               containers::FlatHashMap<CallbackId, DescriptedCallback>>;
+  using Callbacks = containers::FlatHashMap<ServerID, RebootIds>;
+  struct PeerState {
+    std::string serverId;
+    RebootId rebootId{0};
+  };
 
-  CallbackId getNextCallbackId() noexcept;
-
-  void unregisterCallback(PeerState const&, CallbackId);
+  explicit RebootTracker(SchedulerPointer scheduler);
 
-  // bool notifyChange(PeerState const& peerState);
+  CallbackGuard callMeOnChange(PeerState peer, Callback callback,
+                               std::string description);
 
-  void scheduleAllCallbacksFor(ServerID const& serverId);
-  void scheduleCallbacksFor(ServerID const& serverId, RebootId rebootId);
+  void updateServerState(State state);
 
-  static Callback createSchedulerCallback(
-      std::vector<
-          std::shared_ptr<std::unordered_map<CallbackId, DescriptedCallback>>>
-          callbacks);
+ private:
+  void unregisterCallback(PeerState const& peer,
+                          CallbackId callbackId) noexcept;
 
-  void queueCallbacks(
-      std::vector<
-          std::shared_ptr<std::unordered_map<CallbackId, DescriptedCallback>>>
-          callbacks);
-  void queueCallback(DescriptedCallback callback);
+  void queueCallback(DescriptedCallback&& callback) noexcept;
+  void queueCallbacks(std::string_view serverId, RebootId to);
 
- private:
-  mutable Mutex _mutex;
+  mutable std::mutex _mutex;
 
   CallbackId _nextCallbackId{1};
 
+  /// @brief Save a pointer to the scheduler for easier testing
+  SchedulerPointer _scheduler;
+
   /// @brief Last known rebootId of every server.
   /// Will regularly get updates from the agency.
   /// Updates may not be applied if scheduling the affected callbacks fails, so
   /// the scheduling will be tried again on the next update.
-  std::unordered_map<ServerID, RebootId> _rebootIds;
+  State _state;
```

Comment:
Stylistic: I somewhat dislike the name `_state` here, but would prefer to keep the old name `_rebootIds`.
RebootIds was quite a telling variable name, but `State` is not very telling.

Reply:
https://github.com/arangodb/arangodb/pull/16375#discussion_r891388779